Newsletters: add 316 (2024-08-16) (#1821)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2024-08-16-newsletter.md

---
title: 'Bitcoin Optech Newsletter #316'
permalink: /en/newsletters/2024/08/16/
name: 2024-08-16-newsletter
slug: 2024-08-16-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME:harding

## News

<!-- Note: confirmed via email that "Zawy" is how this person would like be attributed.  -harding -->

- **New timewarp vulnerability in testnet4:** Mark "Murch"
  Erhardt [posted][erhardt warp] to Delving Bitcoin to describe an
  attack [discovered][zawy comment] by developer Zawy for exploiting

>   [testnet4][topic testnet]'s new difficulty adjustment algorithm.

Author: I noticed you calling it an alternating timestamp attack and followed suit in my initial rough draft, but after I illustrated your version of the attack and saw how wonky it looked, I thought it deserved the title of being a time warp attack even more than the original time warp attack.  Additionally, Zawy originally described it as a time warp and it meets our short [definition](https://bitcoinops.org/en/topics/time-warp/) of a time warp:

> Time warp is an exploit of Bitcoin’s difficulty adjustment algorithm that allows miners controlling a large amount of hashrate to prevent difficulty from increasing even as the rate of block production increases.

I was also motivated by "alternating timestamp attack" not sounding quite right to me (the timestamps don't alternate, e.g. 1 3 2 4, they jump around) and other people calling the it the Zawy attack and the Murch-Zawy attack, which I found annoying because I find using proper names in technical terms annoying (they lead to random mid-sentence capitalization, often don't credit (all) the right people (e.g. Diffie-Hellman), and sometimes credit someone who was counterproductive (e.g. Claus Schnorr enforcing a patent on schnorr signatures).

It's commonly the case that we initially describe things under a different name than what the development community comes to rough consensus on and we begin using the consensus term later, so I think it's ok if that happens in this case.

  original value to return the rate of blocks being added to the
  mainchain to one block every 10 minutes on average:

  ![Illustration of block censorship by an attacker with slightly more than 50% of total network hashrate (simplified)](/img/posts/2024-time-warp/50p-attack.png)

  A time warp attack occurs when the dishonest miners use their hashrate
  majority to force the timestamps in most blocks to use the minimum
  allowed value.  At the end of each 2,016-block retarget period, they
  increase block header time to the current [wall time][] to make it
  seem like it took longer to produce the blocks than it actually did,
  leading to a lower difficulty in the subsequent period.

  <!-- TODO:harding can probably integrate the illustration below into
  the time warp topic page -->

  ![Illustration of a classic time warp attack (simplified)](/img/posts/2024-time-warp/classic-time-warp.png)

  The [new rule][testnet4 rule] applied to testnet4 fixes this by
  requiring the first block in a new retarget period have a timestamp no
  earlier than 7,200 seconds before its previous block (the last block in
  the previous period).  Note: this was originally proposed as a

>   600-second rule (see [Newsletter #36][news36 warp rule]); additional

Reviewer: https://github.com/bitcoin/bips/pull/1660

>   600-second rule (see [Newsletter #36][news36 warp rule]); additional

Author: Rewrote to a more generic form that doesn't mention the specific number of seconds.

  Fields).  This should not affect anyone who uses the pre-built
  binaries available from BitcoinCore.org---which is what we expect most
  people use.  However, developers and integrators who build their own
  binaries for testing or customization may be affected---especially
  those working on uncommon platforms or build configurations.

  Fields's email provides answers to anticipated questions and asks
  anyone who builds Bitcoin Core themselves to test [PR #30454][bitcoin
  core #30454] and report any issues.  That PR is expected to be merged
  in the next several weeks with anticipated release in version 29
  (anticipated about 7 months from now).  The earlier you test, the more
  time Bitcoin Core developers will have to fix problems before the
  release of version 29---increasing the chance that they can prevent
  problems in the released code from affecting your configuration.

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review
Club][] meeting, highlighting some of the important questions and
answers.  Click on a question below to see a summary of the answer from

> the meeting.*

Reviewer 2: This probably shouldn't be here?

Author: Indeed!  Whoops.  Removed.  Thanks!

  corruption by anti-virus software. This can be optionally disabled and doesn't
  protect against intentional data corruption attacks. This was implemented for
  the `chainstate` files in [Bitcoin Core #6650][] in September 2015.

- [Core Lightning #7528][] adjusts the [fee rate estimation][topic fee
  estimation] for sweeps of non-time-sensitive unilateral channel closures to an
  absolute deadline of 2016 blocks (approximately 2 weeks). Previously, the fee
  rate was set to target confirmation within 300 blocks, which caused
  transactions to be stuck at the minimum relay fee limit, resulting in
  indefinite delays.

- [Core Lightning #7533][] updates the internal coin movement notifications and
  the transaction bookkeeper to properly account for the spending of funding
  outputs for [splicing][topic splicing] transactions. Previously, there was no
  logging or tracking of this.

- [Core Lightning #7517][] askrene

- [LND #8955][] yyforyongyu/cr-8516-240729-sendcoins

- [LND #8886][] bitromortac/buildroute-inbound-fees

- [LND #8967][] add wire messages for quiescence

- [LDK #3215][] tx-sync: Protect against Core's Merkle leaf node weakness

> - [BIPs #1658][] BIP94: Change timewarp delta to 7200 seconds

Reviewer: Note https://github.com/bitcoin/bips/pull/1660

Reviewer 3: if this is getting reverted, I guess we should remove this entry?

Reviewer: Yeah, I think so

---
title: 'Bitcoin Optech Newsletter #316'
permalink: /en/newsletters/2024/08/16/
name: 2024-08-16-newsletter
slug: 2024-08-16-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME:harding

## News

<!-- Note: confirmed via email that "Zawy" is how this person would like be attributed.  -harding -->

- **New timewarp vulnerability in testnet4:** Mark "Murch"
  Erhardt [posted][erhardt warp] to Delving Bitcoin to describe an
  attack [discovered][zawy comment] by developer Zawy for exploiting
  [testnet4][topic testnet]'s new difficulty adjustment algorithm.
  Testnet4 applied a solution from the [consensus cleanup][topic

>   consensus cleanup] soft fork for mainnet that is intended to block the

Reviewer 4: ```suggestion
  consensus cleanup] soft fork for mainnet intended to block the
```

>   [time warp attack][topic time warp].  However, Zawy described an

Reviewer 4: We've used both "time warp" and "timewarp" so maybe this is okay, although I slightly prefer timewarp. Is this a good addition to the [style guide](https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md#compound-words)?

Author: I think the style guide should say something like: "When we have a topic page for a term, prefer the style  dominant on that page unless otherwise mentioned here".  Our topic page is "time warp", so I'll revise the "timewarp" on this newsletter to add the space.

  50% of honest miners.  That would naturally lead initially to only one
  block being produced every 20 minutes on average.  After 2,016 blocks
  are produced using this pattern, difficulty will adjust to 1/2 its
  original value to return the rate of blocks being added to the
  mainchain to one block every 10 minutes on average:

  ![Illustration of block censorship by an attacker with slightly more than 50% of total network hashrate (simplified)](/img/posts/2024-time-warp/50p-attack.png)

  A time warp attack occurs when the dishonest miners use their hashrate
  majority to force the timestamps in most blocks to use the minimum
  allowed value.  At the end of each 2,016-block retarget period, they
  increase block header time to the current [wall time][] to make it
  seem like it took longer to produce the blocks than it actually did,
  leading to a lower difficulty in the subsequent period.

  <!-- TODO:harding can probably integrate the illustration below into
  the time warp topic page -->

  ![Illustration of a classic time warp attack (simplified)](/img/posts/2024-time-warp/classic-time-warp.png)

  The [new rule][testnet4 rule] applied to testnet4 fixes this by

>   requiring the first block in a new retarget period have a timestamp no

Reviewer 4: ```suggestion
  requiring the first block in a new retarget period to have a timestamp no
```

  miners wait for their blocks to be accepted, they can create
  additional blocks at ever lower difficulties.  Any blocks created by
  honest miners during the 16 weeks that attackers are preparing their
  attack will be reorganized out of the chain when full nodes begin
  accepting the attacker blocks; this could mark every transaction
  confirmed during that time as either unconfirmed or invalid
  (_conflicted_) on the current chain.

  Erhardt suggests solving the attack with a soft fork that requires
  the timestamp of the last block in a retarget period be greater than
  the timestamp of the first block in that period.  Zawy proposed
  several solutions, including forbidding block timestamps from
  decreasing (which could create problems if some miners create blocks near
  the two-hour future limit enforced by nodes), or at least forbidding
  them from decreasing by more than about two hours.

  Overall, on mainnet, the new time warp attack is similar to the
  original attack in its requirements for mining equipment, its ability
  to be detected in advance, its consequences for users, and the
  relative simplicity of a soft fork to fix it.  It depends on an

>   attacker maintaining control over at least 50% of hashrate for a
>   period of at least a month, while likely signaling to users that an

Reviewer 4: ```suggestion
  attacker maintaining control over at least 50% of hashrate for
  at least a month, while likely signaling to users that an
```

  attack was impending and hoping they don't respond, which could be
  quite challenging on mainnet.  As Zawy [notes][zawy testnet risk], the
  attack is much easier on testnet: a small amount of modern mining
  equipment is enough to achieve 50% of hashrate on testnet and set up
  the attack in stealth.  An attacker could then, in theory, produce
  over 500,000 blocks per day.  Only someone willing to dedicate a
  greater amount of hashrate to testnet could stop an attacker unless
  the attack is prevented using a soft fork.

  At the time of writing, the tradeoffs between proposed solutions were
  being discussed.

- **Onion message DoS risk discussion:** Gijs van Dam [posted][vandam
  onion] to Delving Bitcoin to discuss a recent [paper][bk onion] by
  researchers Amin Bashiri and Majid Khabbazian about [onion
  messages][topic onion messages].  The researchers note each onion

>   message can be forwarded across a large number of nodes (481 hops by

Reviewer 4: ```suggestion
  message can be forwarded across many nodes (481 hops by
```

  He then describes three potential candidates for an additional opt-in
  mechanism that Bob can use to sign his payments to Alice.  If Bob uses
  that mechanism, Alice's LN wallet can authenticate that
  signature as belonging to Bob and display that information to her.
  In unauthenticated payments, fields set by the spender (such as the
  free-form `payer_note` field) can be marked as untrusted to discourage
  users from relying on information provided in them.

  Feedback on which cryptographic methods to use is requested, with
  Teinturier planning to release [BLIP42][blips #42] with a
  specification for the methods selected.

- **Bitcoin Core switch to CMake build system:** Cory Fields
  [posted][fields cmake] to the Bitcoin-Dev mailing list to announce
  Bitcoin Core's impending switch from the GNU autotools build system to
  the CMake build system, which has been led by Hennadii Stepanov with
  contributions from Michael Ford for bug fixes and modernization, with
  reviews and contributions from several other developers (including
  Fields).  This should not affect anyone who uses the pre-built
  binaries available from BitcoinCore.org---which is what we expect most

>   people use.  However, developers and integrators who build their own

Reviewer 4: ```suggestion
  people are using.  However, developers and integrators who build their own
```

  loading the snapshot and start downloading blocks from that block, defeating
  the purpose of assumeUTXO.

- [Bitcoin Core #30598][] removes block height from the [assumeUTXO][topic
  assumeutxo] snapshot file metadata as it is not a unique identifier in a
  pre-sanitized untrusted file compared to the block hash that is kept. A node
  can still obtain block height from many other internal sources.

- [Bitcoin Core #28280][] optimizes Initial Block Download (IBD) performance for
  pruned nodes by not emptying the UTXO cache during pruning flushes. It does
  this by listing "dirty" cache entries, entries that have changed since they
  were last written to the database, separately. This allows a node to avoid
  unnecessarily scanning the entire cache during prune flushes and instead focus
  on the dirty entries.  The optimization results in up to 32% faster IBD for
  pruned nodes with high `dbcache` settings, and about 9% improvement with
  default settings. See Newsletter [#304][news304 cache].

- [Bitcoin Core #28052][] adds XOR encoding to `blocksdir *.dat` files on
  creation as a preventative mechanism against unintentional and accidental data

>   corruption by anti-virus software or other type of software. This can be

Reviewer 4: ```suggestion
  corruption by anti-virus or similar software. This can be
```
Or, if that's not quite the right meaning, probably should be "other types"
